ledger/balance.go

```diff
@@ -0,0 +1,402 @@
+// Add an Amount to Balance.Balance.
+func (b *Balance) Add(amount Amount) error {
+	result, err := b.Balance.Add(amount)
+	if err != nil {
+		return err
+	}
+	b.Balance = result
+	return nil
+}
+
+// Sub an Amount from Balance.Balance.
+func (b *Balance) Sub(amount Amount) error {
+	result, err := b.Balance.Sub(amount)
+	if err != nil {
+		return err
+	}
+	b.Balance = result
+	return nil
+}
+
+// Lock adds an Amount to Balance.Locked. Returns an error if the result Balance.Locked
+// is greater than the whole Balance.Balance
+func (b *Balance) Lock(amount Amount) error {
+	balance, err := b.Balance.Sub(amount)
+	if err != nil {
+		return err
+	}
+	locked, err := b.Locked.Add(amount)
+	if err != nil {
+		return err
+	}
+	b.Balance = balance
+	b.Locked = locked
+	return nil
+}
+
+// Unlock subs an Amount from Balance.Locked until reach zero.
+func (b *Balance) Unlock(amount Amount) error {
+	locked, subbed, err := b.Locked.SubToZero(amount)
+	if err != nil {
+		return err
+	}
+	balance, err := b.Balance.Add(subbed)
+	if err != nil {
+		return err
+	}
+	b.Balance = balance
+	b.Locked = locked
+	return nil
+}
+
+// Supply an Amount from Balance.Balance to Balance.Supplied.
+// Returns an error if not sufficient Balance.Balance.
+func (b *Balance) Supply(amount Amount) error {
+	balance, err := b.Balance.Sub(amount)
+	if err != nil {
+		return err
+	}
+	supplied, err := b.Supplied.Add(amount)
+	if err != nil {
+		return err
+	}
+	b.Balance = balance
+	b.Supplied = supplied
+	return nil
+}
+
+// Withdraw an Amount from Balance.Supplied to Balance.Balance.
+// Returns an error if not sufficient Balance.Supplied.
+func (b *Balance) Withdraw(amount Amount) error {
+	supplied, err := b.Supplied.Sub(amount)
+	if err != nil {
+		return err
+	}
+	balance, err := b.Balance.Add(amount)
+	if err != nil {
+		return err
+	}
+	b.Balance = balance
+	b.Supplied = supplied
+	return nil
+}
+
+// Borrow an Amount from Leverage module and adds it to Balance.Balance and Balance.Borrowed.
+func (b *Balance) Borrow(amount Amount, isVirtual bool) error {
+	balance, err := b.Balance.Add(amount)
+	if err != nil {
+		return err
+	}
+
+	if isVirtual {
+		virtual, err := b.Virtual.Add(amount)
+		if err != nil {
+			return err
+		}
+		b.Virtual = virtual
+	} else {
+		borrowed, err := b.Borrowed.Add(amount)
+		if err != nil {
+			return err
+		}
+		b.Borrowed = borrowed
+	}
+
+	b.Balance = balance
+
+	return nil
+}
+
+// Repay an Amount from Balance.Borrowed.
+// Returns an error if not sufficient Balance.Borrowed.
+func (b *Balance) Repay(amount Amount) error {
+	borrowed, err := b.Borrowed.Sub(amount)
+	if err != nil {
+		return err
+	}
+	b.Borrowed = borrowed
+	return nil
+}
+
+func (b *Balance) String() string {
+	return fmt.Sprintf(
+		"%s - %s - %s - %s - %s - %s - %s - %s",
+		b.UserID,
+		b.AssetID,
+		b.Balance.String(),
+		b.Borrowed.String(),
+		b.Collateral.String(),
+		b.Supplied.String(),
+		b.Virtual.String(),
+		b.Locked.String(),
+	)
+}
+
+func GetUserBalances(
```

Comment:
Can possibly separate this file into two - one containing the balance methods, and the other containing the DB logic

Reply:
sure